python/shotgun_data/shotgun_data_retriever.py

```diff
@@ -679,22 +787,35 @@ def _get_thumbnail_path(url, bundle, directory_only=False):
         # sharding methodology, see 
         # http://stackoverflow.com/questions/13841931/using-guids-as-folder-names-splitting-up
         #
-        # From the hash, generate paths on the form C1C2/C3C4/rest_of_hash.jpeg
-        # (where C1 is the first character of the hash.)
-        # for a million evenly distributed items, this means ~15 items per folder
+        # From the hash, generate paths on the form C1C2/C3C4/rest_of_hash.ext
+        # (where C1 is the first character of the hash). For a million evenly distributed
+        # items, this means ~15 items per folder
         first_folder = hash_str[0:2]
         second_folder = hash_str[2:4]
-        path_chunks = [first_folder, second_folder]
+
+        # Establish the cache path location
+        cache_path_items = [
+            bundle.cache_location, "thumbs", first_folder, second_folder
+        ]
 
         # If we were only asked to give back a directory path then we can
         # skip building and appending a file name.
         if not directory_only:
-            path_chunks.append("%s.jpeg" % hash_str[4:])
+            path_base = hash_str[4:]
+            url_base, url_ext = os.path.splitext(url_obj.path)
+            if url_ext:
+                cache_path_items.append("%s%s" % (path_base, url_ext))
```

Comment:
wondering if we are are making this too complicated.

maybe it's simpler like this:

```
# Case A: Cache file does not exist

(path, cache_exists) = _get_thumbnail_path("https://foo/bar/baz.jpg") 
# return data (path, cache_exists) = ('/tmp/xx/yy/1245/6678', False) 
# we always get a path without extension back whenever a cache item doesn't exist
# we return False to indicate that a cache file cannot be found
# note: we don't try to deduce any file extension from the url 

# now we therefore download the thumb
full_path = sgtk.util.download_url("https://foo/bar/baz.jpg", path)
# download_url appends the extension and 
# full_path now contains /tmp/xx/yy/1245/6678.jpg

# Case B: cache is warm
(path, cache_exists) = _get_thumbnail_path("https://foo/bar/baz.jpg") 
# return data (path, cache_exists) = ('/tmp/xx/yy/1245/6678.jpg', True)
# all good job done
```

The cache file `/tmp/xx/yy/1245/6678` is unique and generated from the full url - we assume that 
there is only ever one `/tmp/xx/yy/1245/6678.*` file on disk - there cannot be say a `/tmp/xx/yy/1245/6678.jpg` and a `/tmp/xx/yy/1245/6678.png` - all we need to look for is the first match.

This behaviour seems simpler to me. What do you think? 

Also note my comments on the core pull request related to this - i think the download-url method should always **append** the extension just to keep things simple.